### ISSUE TITLE
DiffMask Training/Inference Details

### ISSUE THREAD

**Issue by pedr0sorio** [HINTLab/LeFusion#12]
Hi LeFusion Team,

Thanks again for open-sourcing your interesting work!

I have been trying to reproduce the DiffMask training on my similarly processed version of the LIDC-IDRI dataset and I wonder if you could give additional input on a few points:

#### Model Training
- In order to get the shared weights, how many training steps, effective batch size and GPU setup did you use? Are the configs in `DiffMask/train/config` accurate? I can't seem to find this info in the paper and in my 4xA100 setup this training is about twice as slow as the LeFusion model.
- Did you use any specific metric for checkpoint selection?
- I am getting some weird mask predictions after training a DiffMask model with your code (50k steps with same HPs as in `DiffMask/train/config`) and my preprocessed version of LIDC-IDRI (see below). Did you notice a similar behaviour?
  - <img src="https://github.com/user-attachments/assets/d3ba7898-770d-4be4-85c2-bf9352cdf96f" alt="Image" width="300"/>
  - <img src="https://github.com/user-attachments/assets/ab1166f2-0d4b-43a2-aec6-01aaa1ff99d8" alt="Image" width="300"/>
  - [EDIT] I run the training using your code, your LIDC-IDRI version and `DiffMask/train/config` and I was not able to find a checkpoint that leads to as good predictions as the weights you open-sourced. See below your predictions (top) and my results after 70k steps (bottom) while using the same control spheres as above:
  - <img src="https://github.com/user-attachments/assets/8944c419-6c79-4bae-8795-5e83f159b8c4" alt="Image" width="300"/>
  - <img src="https://github.com/user-attachments/assets/ad293d40-621b-4f7a-b944-e9ba3a8a6809" alt="Image" width="300"/>

- [EDIT] In line:
  - https://github.com/M3DV/LeFusion/blob/eaf2aa27b0aba3e9730615b43f8e96b2dc4fceb1/DiffMask/dataset/lidc_mask.py#L94
  - One of your preprocessing steps includes converting both the sphere mask and the mask from the `[0,1]` range to `[-1,1]` using basic tensor operations. Although, since both the mask & sphere are being loaded as `torch.uint8` (range fixed to `[0,255]`), -1 gets mapped to 255 and we get an inversion of both masks. This does not align with the model weight that you've provided considering that the predictions are most intense in the mask region and thus suggest that model was trained to predict non-inverted / regular masks. To fix this it one could use `torchio` transforms for this rescaling or just convert the masks to `torch.float32` before this operation.

#### Dataset
- Returning to one of the questions raised in #11, could you provide further info on which resolution you train your DiffMask model?
- [EDIT] How do you deal with lesions whose diameter is larger than the fixed context window you've chosen (32,64,64)?
- I notice in `DiffMask/inference/inference.py` you have a final mask processing step which consists of resampling the generated mask to 1.5x finer resolution isotropically: https://github.com/M3DV/LeFusion/blob/eaf2aa27b0aba3e9730615b43f8e96b2dc4fceb1/DiffMask/inference/inference.py#L117
  - Could you comment on how this last step can be useful for training the downstream segmentation model if the crop size is fixed to (32, 64, 64)? 
  
Thanks in advance and keep up with the great work!

**Comment by kanydao** [comment]
@pedr0sorio 
Thank you for your attention!

For mask training, since the masks are binary (0 and 1), we needed to normalize them to the range of [-1, 1] to align with the diffusion model's paradigm. Therefore, we applied the transformation mask = mask * 2 - 1, which involves floating-point operations. I suspect that the issues encountered may be related to differences in hardware or software versions.

During training, we found that mask diffusion is quite sensitive. We recommend using a small learning rate and a relatively low number of epochs, as the model converges very quickly. In our experience, convergence typically occurs within 10 epochs (around 7000 iterations usually gives good results). Overtraining may lead to collapse, resulting in fragmented or abnormal masks, which we addressed through post-processing techniques.

In our experiments, we also found that to achieve more stable mask training, encoding the masks using Signed Distance Functions (SDFs) is an effective and theoretically sound approach. We are currently conducting further research in this area, and related work will be released in the future.

As for the mask size, we did not specifically focus on this aspect. Our experiments involved lesions generally within the window. However, if larger masks are needed, a sliding window approach could be used. Within the window, different sphere sizes can be employed to control the mask size, but they should not exceed the window boundaries.

Hantao Zhang
